processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java

```diff
@@ -19,66 +19,266 @@
+import io.druid.segment.data.CompressedVSizeIntsIndexedSupplier;
 import io.druid.segment.data.GenericIndexed;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.IndexedMultivalue;
 import io.druid.segment.data.IndexedRTree;
 import io.druid.segment.data.VSizeIndexed;
 import io.druid.segment.data.VSizeIndexedInts;
+import io.druid.segment.data.WritableSupplier;
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
 import java.nio.channels.WritableByteChannel;
 
-/**
- */
+
 public class DictionaryEncodedColumnPartSerde implements ColumnPartSerde
 {
-  private final boolean isSingleValued;
+  private static final int NO_FLAGS = 0;
+
+  enum Feature {
+    MULTI_VALUE;
+    public boolean isSet(int flags) { return (getMask() & flags) != 0; }
+    public int getMask() { return (1 << ordinal()); }
+  }
+
+  enum VERSION
+  {
+    UNCOMPRESSED_SINGLE_VALUE,  // 0x0
+    UNCOMPRESSED_MULTI_VALUE,   // 0x1
+    COMPRESSED;                 // 0x2
+
+    public static VERSION fromByte(byte b) {
+      final VERSION[] values = VERSION.values();
+      Preconditions.checkArgument(b < values.length, "Unsupported dictionary column version[%s]", b);
+      return values[b];
+    }
+
+    public byte asByte() {
+      return (byte)this.ordinal();
+    }
+  }
+
+  public static class Builder {
+    private GenericIndexed<String> dictionary = null;
+    private VSizeIndexedInts singleValuedColumn = null;
+    private CompressedVSizeIntsIndexedSupplier compressedSingleValuedColumn = null;
+    private VSizeIndexed multiValuedColumn = null;
+    private BitmapSerdeFactory bitmapSerdeFactory = null;
+    private GenericIndexed<ImmutableBitmap> bitmaps = null;
+    private ImmutableRTree spatialIndex = null;
+    private ByteOrder byteOrder = null;
+
+    private Builder()
+    {
+    }
+
+    public Builder withDictionary(GenericIndexed<String> dictionary)
+    {
+      this.dictionary = dictionary;
+      return this;
+    }
+
+    public Builder withBitmapSerdeFactory(BitmapSerdeFactory bitmapSerdeFactory)
+    {
+      this.bitmapSerdeFactory = bitmapSerdeFactory;
+      return this;
+    }
+
+    public Builder withBitmaps(GenericIndexed<ImmutableBitmap> bitmaps)
+    {
+      this.bitmaps = bitmaps;
+      return this;
+    }
+
+    public Builder withSpatialIndex(ImmutableRTree spatialIndex)
+    {
+      this.spatialIndex = spatialIndex;
+      return this;
+    }
+
+    public Builder withByteOrder(ByteOrder byteOrder)
+    {
+      this.byteOrder = byteOrder;
+      return this;
+    }
+
+    public Builder withSingleValuedColumn(VSizeIndexedInts singleValuedColumn)
+    {
+      Preconditions.checkState(compressedSingleValuedColumn == null, "Cannot set both singleValuedColumn and compressedSingleValuedColumn");
+      Preconditions.checkState(multiValuedColumn == null, "Cannot set both singleValuedColumn and multiValuedColumn");
+      this.singleValuedColumn = singleValuedColumn;
```

Comment:
Why wouldn't this do the `.asWriteableSupplier()` and simplify the build() call to just one thing?